src/Runtime/Distributions/Automata/Automaton.Builder.cs

```diff
@@ -0,0 +1,810 @@
+                        this.builder.transitions[this.index] = node;
+                    }
+                }
+
+                /// <summary>
+                /// Marks current transition as removed. This transition will not be visible during further
+                /// iterations.
+                /// </summary>
+                /// <remarks>
+                /// <see cref="Remove()"/> can be called only once until <see cref="Next()"/> call.
+                /// </remarks>
+                public void Remove()
+                {
+                    var state = this.builder.states[this.stateIndex];
+                    var node = this.builder.transitions[this.index];
+
+                    // unlink from previous node
+                    if (node.Prev != -1)
+                    {
+                        var prevNode = builder.transitions[node.Prev];
+                        prevNode.Next = node.Next;
+                        builder.transitions[node.Prev] = prevNode;
+                    }
+
+                    // unlink from next node
+                    if (node.Next != -1)
+                    {
+                        var nextNode = builder.transitions[node.Next];
+                        nextNode.Prev = node.Prev;
+                        builder.transitions[node.Next] = nextNode;
+                    }
+
+                    // update references in state
+                    if (state.FirstTransition == this.index || state.LastTransition == this.index)
+                    {
+                        if (state.FirstTransition == this.index)
+                        {
+                            state.FirstTransition = node.Next;
+                        }
+
+                        if (state.LastTransition == this.index)
+                        {
+                            state.LastTransition = node.Prev;
+                        }
+
+                        this.builder.states[this.stateIndex] = state;
+                    }
+
+                    ++this.builder.numRemovedTransitions;
+                }
+
+                /// <summary>
+                /// Gets a value indicating whether iterator is dereferenceable - its Value can be get or set.
+                /// Once iteration is finished property will become false.
+                /// </summary>
+                public bool Ok => this.index != -1;
+
+                /// <summary>
+                /// Moves iterator to next transition in list.
+                /// </summary>
+                public void Next()
+                {
+                    this.index = this.builder.transitions[this.index].Next;
+                }
+
+            }
+
+            /// <summary>
+            /// Wrapper aro
+            /// </summary>
+            private struct LinkedTransitionNode
+            {
+                /// <summary>
+                /// Stored transition.
+                /// </summary>
+                public Transition Transition;
```

Comment:
Structs should be readonly, is there a reason this one is not?

Reply:
Well, basically all this exercise is about fitting data into array of structures. Which requires mutable data during construction.

But in this specific case it actually can be readonly. Will fix

Reply:
No, it can't. Will add a comment why.

Reply:
Added comment in struct <summary>.